[expo-in-app-purchases] Add defensive null checks and missing iOS SKErrorCode values

danmaas opened the pull request:
# Why

We have seen several crash reports in production on Android and iOS due to the platform IAP code sending invalid or missing Product SKU data to `expo-in-app-purchases`.

On Android, despite the documentation implying that the parameters to methods like `onQueryPurchasesFinished()` and `onSkuDetailsResponse()` are not nullable, we have seen occasional crashes in production where they have been null. This patch adds some defensive null checks to the Android code to gracefully ignore invalid input rather than crash.

On iOS, there are cases when some fields of `SKProduct` coming from the StoreKit API can be `nil`, specifically during the app review process when your IAP product is in a "Review Rejected" state. In this case, I think it's best to ignore the invalid product SKU, otherwise the `nil` values will propagate through the EXInAppPurchases code and inevitably cause a native Objective-C crash.

(example crash report: https://sentry.io/share/issue/057be1d8c2f345c687feb82349a87adb/ - note, I believe the part of the stack trace that includes `expo::gl_cpp` is bogus due to threading or JSC issues - the actual error comes from receiving a SKProduct containing `nil` values from `[SKProductsRequest _start]` and passing it to `getProductData`)

In addition, this patch adds a few missing `SKErrorCode` values to `errorCodeNativeToJS` that were added in more recent StoreKit versions. (happy to separate this part if you'd like).

By the way, separately, I'd like to propose updating the Android Google Play Billing SDK to 2.2.0, which is the latest version that has no breaking changes relative to the version currently used by `expo-in-app-purchases` (2.0.0).

# How

Ignore invalid product SKUs at the Android/iOS API level.

# Test Plan

We've been running this patch in production on our app (https://badukpop.com) for several months with satisfactory results.

danmaas committed:
Add defensive null checks and missing iOS SKErrorCode values

danmaas committed:
fix comment typos

ide commented on packages/expo-in-app-purchases/android/src/main/java/expo/modules/inapppurchases/BillingManager.java:
```diff
@@ -501,9 +501,11 @@ public void queryPurchasableItems(List<String> itemList, final Promise promise)
-            mSkuDetailsMap.put(skuDetails.getSku(), skuDetails);
-            results.add(skuToBundle(skuDetails));
+          if (skuDetailsList != null) {
+              for (SkuDetails skuDetails : skuDetailsList) {
+                mSkuDetailsMap.put(skuDetails.getSku(), skuDetails);
+                results.add(skuToBundle(skuDetails));
+              }
```
```suggestion
            for (SkuDetails skuDetails : skuDetailsList) {
              mSkuDetailsMap.put(skuDetails.getSku(), skuDetails);
              results.add(skuToBundle(skuDetails));
            }
```

ide commented on packages/expo-in-app-purchases/ios/EXInAppPurchases/EXInAppPurchasesModule.m:
```diff
@@ -151,6 +151,7 @@ - (void)handleQuery:(SKProductsResponse *)response {
   NSMutableArray *result = [NSMutableArray array];
   
   for (SKProduct *validProduct in response.products) {
+    if(!validProduct.localizedDescription) { continue; } // skip product with nil values - this can happen if it is in review "rejected" state
```
```suggestion
    if (!validProduct.localizedDescription) { continue; } // skip product with nil values - this can happen if it is in review "rejected" state
```

ide commented on packages/expo-in-app-purchases/ios/EXInAppPurchases/EXInAppPurchasesModule.m:
```diff
@@ -167,6 +168,7 @@ -(void)handlePurchase:(SKProductsResponse *)response {
   }
   
   for (SKProduct *validProduct in response.products) {
+    if(!validProduct.localizedDescription) { continue; } // skip product with nil values - this can happen if it is in review "rejected" state
```
```suggestion
    if (!validProduct.localizedDescription) { continue; } // skip product with nil values - this can happen if it is in review "rejected" state
```

ide commented on packages/expo-in-app-purchases/ios/EXInAppPurchases/EXInAppPurchasesModule.m:
```diff
@@ -422,12 +424,14 @@ - (int)errorCodeNativeToJS:(SKErrorCode)errorCode
     case SKErrorCloudServiceRevoked:
     case SKErrorCloudServicePermissionDenied:
     case SKErrorCloudServiceNetworkConnectionFailed:
+    case SKErrorOverlayTimeout:
```
Instead of "CLOUD_SERVICE" as the error, it seems like "SERVICE_TIMEOUT" (4) would be a closer match.

ide commented on packages/expo-in-app-purchases/ios/EXInAppPurchases/EXInAppPurchasesModule.m:
```diff
@@ -436,8 +440,11 @@ - (int)errorCodeNativeToJS:(SKErrorCode)errorCode
+    case SKErrorIneligibleForOffer:
       return 13;
     case SKErrorMissingOfferParams:
+    case SKErrorUnsupportedPlatform:
```
Do you know what triggers this? Is it e.g. Catalyst code on macOS for an app that supports IAPs only on iOS?

danmaas replied:
Sorry, I don't know what triggers `SKErrorUnsupportedPlatform`. The Apple docs are empty (https://developer.apple.com/documentation/storekit/skerrorcode/skerrorunsupportedplatform).

I will reassign this to 0 (UNKNOWN) for now. (I've also commented the magic numbers to show which IAPErrorCode enum value they correspond to).

danmaas committed:
Apply suggestions from code review

- BillingManager.java: Formatting updates
- EXInAppPurchasesModule.m: Formatting updates and IAPErrorCode re-assignments

ide commented:
This is on npm as version 10.1.0.

danmaas commented:
Thank you for the speedy review!
